### PR TITLE
Offline DQM for BTagMu and PFJet paths

### PR DESCRIPTION
**Pull request by defranchis**
Both for AK4 and AK8. This PR is based on PR#18981 (TopMonitor)

### CODE REVIEW

**Comment by mtosi** [comment]
thanks !
am I wrong or the b-tagging discriminator is not monitored ?!?!??
I think we need a turnON as function of the b-tag discriminator ....

**Comment by fwyzard** [comment]
tracked by #19142

**Comment by mtosi** [comment]
when will you add the b-discriminator ?

**Comment by defranchis** [comment]
@mtosi I will add it most likely today

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by mtosi** [comment]
@dmitrijus if this is fine, ca you sign it, please ?
thanks !

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by dmitrijus** [comment]
+1

**Comment by defranchis** [comment]
@fwyzard can we please test again?

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by defranchis** [comment]
@dmitrijus could you please check and sign again?

**Comment by dmitrijus** [comment]
+1

**Comment by davidlange6** [comment]
hi @dmitrijus @mtosi -seems there is a web of PRs here. Which one should be reviewed first (then the others rebased)? Otherwise, all of them need to provide an estimate of how many histogram bins are introduced.

**Comment by mtosi** [comment]
either 
https://github.com/cms-sw/cmssw/pull/19119
or
https://github.com/cms-sw/cmssw/pull/18968

should be integrated first,
then many other ones depend on them

File: DQMOffline/Trigger/plugins/TopMonitor.cc [hunk 1]
```diff
@@ -0,0 +1,1024 @@
...
+    ele1Eta_ele2Eta_ = empty;
+    mu1Pt_mu2Pt_ = empty;
+    mu1Eta_mu2Eta_ = empty;
+    elePt_muPt_ = empty;
+    eleEta_muEta_ = empty;
+
+    //BTV
+    DeltaR_jet_Mu_ = empty;
+
+    for (unsigned int iMu=0; iMu<nmuons_; ++iMu){
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: reserve the needed size of these vectors

File: DQMOffline/Trigger/plugins/TopMonitor.h [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  std::vector<METME> muEtaPhi_;
+  // Marina
+  std::vector<METME> bjetPtEta_;
+  std::vector<METME> bjetEtaPhi_;
+  std::vector<METME> bjetCSVHT_;
+
+  METME eventHT_;
+  METME eventHT_variableBinning_;
+
+  GenericTriggerEventFlag* num_genTriggerEventFlag_;
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/TopMonitor.h]: these need to be declared as unique_ptrs

File: DQMOffline/Trigger/plugins/TopMonitor.cc [hunk 3]
```diff
@@ -0,0 +1,1024 @@
...
+TopMonitor::~TopMonitor()
+{
+  if (num_genTriggerEventFlag_) delete num_genTriggerEventFlag_;
+  if (den_genTriggerEventFlag_) delete den_genTriggerEventFlag_;
+}
+
+MEbinning TopMonitor::getHistoPSet(edm::ParameterSet pset)
+{
+  return MEbinning{
+    pset.getParameter<int32_t>("nbins"),
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: this is better as an unsigned int (no concept of negative bins)

File: DQMOffline/Trigger/plugins/TopMonitor.cc [hunk 4]
```diff
@@ -0,0 +1,1024 @@
...
+
+      histname = "muEtaPhi_"; histtitle = "muon #eta - #phi - ";
+      histname.append(index); histtitle.append(index);
+      bookME(ibooker,muEtaPhi_.at(iMu),histname,histtitle, muEta_variable_binning_2D_, phi_variable_binning_2D_);
+      setMETitle(muEtaPhi_.at(iMu),"muon p_{T} [GeV]","muon #eta");
+
+
+  }
+
+  if ((elePt_.size()!=nelectrons_) || (eleEta_.size()!=nelectrons_) || (elePhi_.size()!=nelectrons_)){
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: this is impossible, right?

File: DQMOffline/Trigger/plugins/TopMonitor.cc [hunk 5]
```diff
@@ -0,0 +1,1024 @@
...
+      if ( HTdefinition_ ( j ) ){
+          eventHT += j.pt();
+      }
+      if ( jetSelection_( j ) ){
+          bool isJetOverlappedWithLepton = false;
+          TLorentzVector jp4; jp4.SetPtEtaPhiE(j.pt(),j.eta(),j.phi(),j.energy());
+          if(nmuons_>0){
+              for (auto const m : muons){
+                  TLorentzVector mp4; mp4.SetPtEtaPhiE(m.pt(),m.eta(),m.phi(),m.energy());
+                  if (mp4.DeltaR(jp4)<leptJetDeltaRmin_){
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: use DataFormats/Math/include/deltaR and avoid the either conversion to TLorentzVector

File: DQMOffline/Trigger/plugins/TopMonitor.h [hunk 6]
```diff
@@ -0,0 +1,255 @@
+#ifndef TOPMONITOR_H
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/TopMonitor.h]: please add the package name as part of the include guard

**Comment by davidlange6** [comment]
otherwise how many histogram bins are added?

**Comment by davidlange6** [comment]
hi @mtosi - could you provide developers a base class that 1) avoids most of the copied code from the various trigger dqm developers, and 2) doesn't have the errors all these PRs have? I'm a bit sick of reviewing the same bugs..

**Comment by mtosi** [comment]
I saw your comments, indeed ... and I was thinking the same

it was the case, actually
and unfortunately that code has all the mentioned errors, 
which were not spotted during its integration review :(

I think some of the developers are playing w/ the mentioned template class
as soon as this storm of PRs will be integrated, I'll fix the missing code, if it is fine w/ you

I will try to catch them when reviewing the PRs

**Comment by defranchis** [comment]
@davidlange6 as @mtosi mention, most of the developments are based on the METMonitor class. I will address all you comments now. Thanks a lot for reviewing this PR

File: DQMOffline/Trigger/plugins/TopMonitor.cc (hunk 1 above)
Comment (defranchis) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: do you mean something like this?
muPhi_ = std::vector<METME>(nmuons_,empty)

File: DQMOffline/Trigger/plugins/TopMonitor.h (hunk 2 above)
Comment (defranchis) [inline DQMOffline/Trigger/plugins/TopMonitor.h]: OK

File: DQMOffline/Trigger/plugins/TopMonitor.cc (hunk 3 above)
Comment (defranchis) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: sure

File: DQMOffline/Trigger/plugins/TopMonitor.cc (hunk 4 above)
Comment (defranchis) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: yes, those lines were there only as sanity checks for debugging purposes, I forgot to remove them

File: DQMOffline/Trigger/plugins/TopMonitor.cc (hunk 5 above)
Comment (defranchis) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: OK, thanks

File: DQMOffline/Trigger/plugins/TopMonitor.h (hunk 6 above)
Comment (defranchis) [inline DQMOffline/Trigger/plugins/TopMonitor.h]: do you mean something like
#ifndef DQMOFFLINE_TRIGGER_TOPMONITOR_H etc?

File: DQMOffline/Trigger/plugins/TopMonitor.cc (hunk 4 above)
Comment (fwyzard) [inline DQMOffline/Trigger/plugins/TopMonitor.cc]: if you want to keep them for debugging, you can use an `assert`

File: DQMOffline/Trigger/plugins/TopMonitor.h (hunk 6 above)
Comment (fwyzard) [inline DQMOffline/Trigger/plugins/TopMonitor.h]: yes, either that or DQMOffline_Trigger_TopMonitor_h

**Comment by davidlange6** [comment]
but just above there is a loop creating elePt_ using nelectrons_ (and similarly for the other variables) right? this is not protecting against bad input data (or anything else that I see)

**Comment by defranchis** [comment]
@davidlange6 please let me know if the last commit covers all the issues. @fwyzard I added the missing monitors, can we please test again?

**Comment by mtosi** [comment]
thanks @defranchis 
I've just updated the TSG google doc,
hope they will be integrated in the online menu soon

**Comment by defranchis** [comment]
@fwyzard can we please test again?

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by defranchis** [comment]
@dmitrijus @davidlange6 could you please check again and sign?

**Comment by dmitrijus** [comment]
+1

**Comment by bhawnasinp** [comment]
Hi @defranchis 

I was cloning your request for one of the Wprime paths and I noticed one issue inthe TopMonitor_cfi.py
L97-99, here instead of "num", it should be "den"
hltTOPmonitoring.numGenericTriggerEventPSet.andOrHlt      = cms.bool(True)# True:=OR; False:=AND
hltTOPmonitoring.numGenericTriggerEventPSet.hltInputTag   = cms.InputTag( "TriggerResults::HLT" )
hltTOPmonitoring.numGenericTriggerEventPSet.errorReplyHlt = cms.bool( False )

Could you please cross-check this?

**Comment by defranchis** [comment]
Hi @bhawnasinp you're absolutely right, many thanks for spotting this

**Comment by defranchis** [comment]
@fwyzard can we please check once more?

**Comment by mtosi** [comment]
@cmsbuild, please test

**Comment by dmitrijus** [comment]
+1

**Comment by davidlange6** [comment]
hi @defranchis - thanks for all the answers - I think I just miss how many bins are added by this PR. thx.

File: DQMOffline/Trigger/python/TopMonitoring_cff.py [hunk 7]
```diff
@@ -1,4 +1,577 @@
...
+#DZ monitor
+topElecMuonHLTMonitor_Dz_Mu12Ele23 = topElecMuonHLTMonitor.clone()
+topElecMuonHLTMonitor_Dz_Mu12Ele23.FolderName = cms.string('HLT/TopHLTOffline/TopMonitor/DiLepton/ElecMuon/Mu12Ele23_DzEfficiency/')
+topElecMuonHLTMonitor_Dz_Mu12Ele23.numGenericTriggerEventPSet.hltPaths = cms.vstring('HLT_Mu12_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_DZ_v*')
+topElecMuonHLTMonitor_Dz_Mu12Ele23.denGenericTriggerEventPSet.hltPaths = cms.vstring('HLT_Mu12_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_v*')
+
+topElecMuonHLTMonitor_Dz_Mu8Ele23 = topElecMuonHLTMonitor.clone()
+topElecMuonHLTMonitor_Dz_Mu8Ele23.FolderName = cms.string('HLT/TopHLTOffline/TopMonitor/DiLepton/ElecMuon/Mu8Ele23_DzEfficiency/')
+topElecMuonHLTMonitor_Dz_Mu8Ele23.numGenericTriggerEventPSet.hltPaths = cms.vstring('HLT_Mu8_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_DZ_v*')
+topElecMuonHLTMonitor_Dz_Mu8Ele23.denGenericTriggerEventPSet.hltPaths = cms.vstring('HLT_Mu8_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_v*')
```
Comment (mtosi) [inline DQMOffline/Trigger/python/TopMonitoring_cff.py]: still missing the monitoring of this path !!!!!

Comment (mtosi) [inline DQMOffline/Trigger/python/TopMonitoring_cff.py]: it means, that this path does not go online,
and it invalidate the monitoring of the HLT_Mu8_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_DZ_v !!!!!!!!

File: DQMOffline/Trigger/python/TopMonitoring_cff.py [hunk 8]
```diff
@@ -1,4 +1,577 @@
...
+topElecMuonHLTMonitor.muoSelection = cms.string('pt>15 & abs(eta)<2.4 & (pfIsolationR04.sumChargedHadronPt + max(pfIsolationR04.sumNeutralHadronEt + pfIsolationR04.sumPhotonEt - (pfIsolationR04.sumPUPt)/2.,0.))/pt < 0.25  & isPFMuon & (isTrackerMuon || isGlobalMuon)') 
+topElecMuonHLTMonitor.jetSelection = cms.string('pt>30 & abs(eta)<2.4')
+topElecMuonHLTMonitor.numGenericTriggerEventPSet.hltPaths = cms.vstring(['HLT_Mu8_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_DZ_v*','HLT_Mu12_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_DZ_v*','HLT_Mu23_TrkIsoVVL_Ele12_CaloIdL_TrackIdL_IsoVL_DZ_v*'])
+
+#DZ monitor
+topElecMuonHLTMonitor_Dz_Mu12Ele23 = topElecMuonHLTMonitor.clone()
+topElecMuonHLTMonitor_Dz_Mu12Ele23.FolderName = cms.string('HLT/TopHLTOffline/TopMonitor/DiLepton/ElecMuon/Mu12Ele23_DzEfficiency/')
+topElecMuonHLTMonitor_Dz_Mu12Ele23.numGenericTriggerEventPSet.hltPaths = cms.vstring('HLT_Mu12_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_DZ_v*')
+topElecMuonHLTMonitor_Dz_Mu12Ele23.denGenericTriggerEventPSet.hltPaths = cms.vstring('HLT_Mu12_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_v*')
```
Comment (mtosi) [inline DQMOffline/Trigger/python/TopMonitoring_cff.py]: still missing the monitoring of this path !!!!
=> this path does not go online,
and it invalidate the monitoring of the HLT_Mu12_TrkIsoVVL_Ele23_CaloIdL_TrackIdL_IsoVL_DZ_v !!!!!!!!!!

File: DQMOffline/Trigger/python/TopMonitoring_cff.py (hunk 7 above)
Comment (defranchis) [inline DQMOffline/Trigger/python/TopMonitoring_cff.py]: I am going to add it in a new PR

File: DQMOffline/Trigger/python/TopMonitoring_cff.py (hunk 8 above)
Comment (defranchis) [inline DQMOffline/Trigger/python/TopMonitoring_cff.py]: as above

**Comment by defranchis** [comment]
@davidlange6 is there a smart way to get that information? Thanks

**Comment by davidlange6** [comment]
maybe @dmitrijus or other dqm person does.

File: DQMOffline/Trigger/python/TopMonitoring_cff.py (hunk 7 above)
Comment (defranchis) [inline DQMOffline/Trigger/python/TopMonitoring_cff.py]: added in https://github.com/cms-sw/cmssw/pull/19471

File: DQMOffline/Trigger/python/TopMonitoring_cff.py (hunk 8 above)
Comment (defranchis) [inline DQMOffline/Trigger/python/TopMonitoring_cff.py]: added in https://github.com/cms-sw/cmssw/pull/19471

**Comment by mtosi** [comment]
I'm confused :(
is this still valid ?

**Comment by defranchis** [comment]
@mtosi yes, this is ready to be merged (after I provide the info about the number of bins added). The other one contains some updates. I guess it will be merged after

**Comment by fwyzard** [comment]
@defranchis this PR has automatically been moved to CMSSW 9.3.x .
Can you make a new PR for CMSSW 9.2.x ?
I think at the moment you can use the same branch.

**Comment by defranchis** [comment]
@fwyzard here it is https://github.com/cms-sw/cmssw/pull/19505

**Comment by gomber** [comment]
@defranchis can you provide nbins inforamtion, so that this PR can be merged? I guess thats the only action item left here. There are other PR in exotica also which relies on this one, so it would be helpful if this gets merged!

**Comment by defranchis** [comment]
@gomber is there a smart way to get this information? Or do I have to create my own script?

**Comment by gomber** [comment]
@defranchis Unfortunately I didn't find any script or automatic way to do this. And that's what I was waiting for last one week atleast for my other PR. So at the end in order to just get this done, I spend 5 minutes with my code looking at the number of bins I am using. Just counted them!

**Comment by lathomas** [comment]
Hi, I'm not sure this is the right place to ask but would it be possible to also monitor 
HLT_Mu17_TrkIsoVVL_Mu8_TrkIsoVVL_DZ_Mass3p8 in this DQM?
(since this code is already monitorng HLT_Mu17_TrkIsoVVL_Mu8_TrkIsoVVL_DZ_Mass8 it should be ~trivial?) 
 
Thanks.

**Comment by fwyzard** [comment]
Hi @lathomas , 
given how long it takes for PRs to be reviewed and integrated, please open a new one rather than modify this one, that then will have to be signed all over again.

Ciao,
.Andrea

**Comment by fwyzard** [comment]
@davidlange6 what is holding up this PR from being integrated ?
It has been signed by DQM over two weeks ago.

**Comment by davidlange6** [comment]
from the thread, it looks like only knowing how many bins there are coming in (as was asked _3_ weeks ago...), but I've not looked at this PR for a while

**Comment by defranchis** [comment]
@davidlange6 this PR adds 466424 bins

File: DQMOffline/Trigger/plugins/TopMonitor.h [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+
+class GenericTriggerEventFlag;
+
+struct MEbinning {
+  unsigned int nbins;
+  double xmin;
+  double xmax;
+};
+
+struct METME {
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/TopMonitor.h]: hi @defranchis - looks like I missed three things initially - please take care of these in a followup PR (and as part of the back port)

1) move the definitions of MEbinning, METME and PVcut into the TopMonitor class itself
2) initialize numerator and denominator to nullptr here (then you can remove a bunch of initializations later
3) having something called METME in a TopMonitor class makes no sense. Please rename

**Comment by davidlange6** [comment]
+1

File: DQMOffline/Trigger/plugins/TopMonitor.h (hunk 9 above)
Comment (defranchis) [inline DQMOffline/Trigger/plugins/TopMonitor.h]: OK thanks, I will make a new PR then

**Pull request merged**